### PR TITLE
[CLIPPY] Update templates to allow generated scaffolding to pass pedantic clippy

### PR DESCRIPTION
**Pull request by stan-irl**
## Background
- https://github.com/mozilla/uniffi-rs/issues/1476
- The problem here is that rust scaffolding has to be directly included into the module where the implementation is defined or else is it wont compile. This means that a consumer's code cant have stricter clippy rules than the generated scaffolding.
- Theres not really a way to deal with this except to make the generated scaffolding pass all possible clippy rules
- I think this is fair because consumers of uniffi - I dont have control over the generated code so it should have no impact over their code formatting

## Changes
- Add exceptions of fix all clippy errors that fired with 
```
cargo clippy -- --all-targets --all-features -- \
	-D warnings \
	-D clippy::all \
	-D missing_docs \
	-D clippy::missing_docs_in_private_items \
	-D clippy::pedantic \
	-W clippy::nursery \
	-W clippy::cargo
```
- Add some exceptions so that uniffi scaffolding isnt included in generated cargo docs for consumers

## Tests
- [x] Ingested into my own project and verified by hand
- [x] All uniffi tests pass

### CODE REVIEW

**Comment by stan-irl** [comment]
@mhammond FYI. I think we've just gotta fix these on a case-by-case basis

File: uniffi_bindgen/src/scaffolding/templates/macros.rs [hunk 1]
```diff
@@ -112,7 +112,8 @@ uniffi::call_with_output(call_status, || {
     {% else -%}
     {% if func.full_arguments().is_empty() %}#[allow(clippy::redundant_closure)]{% endif %}
     {% call to_rs_call(func) %}
+    {%- if func.return_type().is_none() %};{% endif %}
     {% endmatch -%}
-})
+}){%- if func.return_type().is_none() -%};{% endif %}
```
Comment (stan-irl) [inline uniffi_bindgen/src/scaffolding/templates/macros.rs:117]: add semicolons for funcs that dont return anything

File: uniffi_bindgen/src/scaffolding/templates/RustBuffer.rs [hunk 2]
```diff
@@ -2,25 +2,29 @@
...
+#[doc(hidden)]
 #[no_mangle]
 pub unsafe extern "C" fn {{ ci.ffi_rustbuffer_free().name() }}(buf: uniffi::RustBuffer, call_status: &mut uniffi::RustCallStatus) {
-    uniffi::ffi::uniffi_rustbuffer_free(buf, call_status)
+    uniffi::ffi::uniffi_rustbuffer_free(buf, call_status);
 }
 
-#[allow(clippy::missing_safety_doc)]
+#[allow(clippy::missing_safety_doc, missing_docs)]
+#[doc(hidden)]
```
Comment (stan-irl) [inline uniffi_bindgen/src/scaffolding/templates/RustBuffer.rs:27]: lets keep this out of consumer cargo docs and also allow them to be undocumented

File: uniffi_bindgen/src/scaffolding/templates/ReexportUniFFIScaffolding.rs [hunk 3]
```diff
@@ -11,10 +11,11 @@
 // megazord).  The combined library has a cargo dependency for each component and calls
 // uniffi_reexport_scaffolding!() for each one.
 
+#[allow(missing_docs)]
 #[doc(hidden)]
-pub fn uniffi_reexport_hack() {
-}
+pub const fn uniffi_reexport_hack() {}
```
Comment (stan-irl) [inline uniffi_bindgen/src/scaffolding/templates/ReexportUniFFIScaffolding.rs:16]: this should be `const` and  we also want to allow it to be undocumented

**Review by mhammond — APPROVED** [review]
seems fine to me, thanks!

**Pull request merged**